File: sdk/ios/DooPushSDKExample-ObjC/DooPushSDKExample-ObjC/PushNotificationManager.m
//
//  PushNotificationManager.m
//  DooPushSDKExample-ObjC
//
//  推送通知管理器 - 实现DooPushDelegate
//

#import "PushNotificationManager.h"
#import "Logger.h"
@import DooPushSDK;

// 状态更新通知常量
NSString * const DooPushStatusUpdateNotification = @"DooPushStatusUpdateNotification";

#pragma mark - NotificationInfo Implementation

@implementation NotificationInfo

- (instancetype)init {
    self = [super init];
    if (self) {
        _identifier = [[NSUUID UUID] UUIDString];
        _receivedAt = [NSDate date];
    }
    return self;
}

- (BOOL)isEqual:(id)object {
    if (![object isKindOfClass:[NotificationInfo class]]) {
        return NO;
    }
    NotificationInfo *other = (NotificationInfo *)object;
    return [self.identifier isEqualToString:other.identifier];
}

- (NSUInteger)hash {
    return [self.identifier hash];
}

@end

#pragma mark - PushNotificationManager Implementation

@interface PushNotificationManager ()

@property (nonatomic, strong) NSTimer *updateTimeoutTimer;

@end

@implementation PushNotificationManager

static PushNotificationManager *_shared = nil;

+ (PushNotificationManager *)shared {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _shared = [[PushNotificationManager alloc] init];
    });
    return _shared;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _notifications = [[NSMutableArray alloc] init];
        _sdkStatus = SDKStatusNotConfigured;
        _pushPermissionStatus = UNAuthorizationStatusNotDetermined;
        _isLoading = NO;
        _isUpdatingDevice = NO;
        
        [self checkPermissionStatus];
        [self updateSDKStatus];
    }
    return self;
}

#pragma mark - Public Methods

- (void)checkAutoRegister {
    // 检查是否之前已经注册过
    NSString *localToken = [DooPushManager.shared getDeviceToken];
    if (localToken && localToken.length > 0) {
        // 如果有本地token，自动重新注册以恢复连接
        [self registerForPushNotifications];
    }
}

- (void)registerForPushNotifications {
    self.isLoading = YES;
    self.sdkStatus = SDKStatusRegistering;
    self.lastError = nil;
    [self notifyStatusUpdate];
    
    __weak typeof(self) weakSelf = self;
    [DooPushManager.shared registerForPushNotificationsWithCompletion:^(NSString * _Nullable token, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            weakSelf.isLoading = NO;
            
            if (error) {
                [weakSelf handleRegistrationError:error];
            } else if (token) {
                [weakSelf handleRegistrationSuccess:token];
            }
        });
    }];
}

- (void)updateDeviceInfo {
    self.isUpdatingDevice = YES;
    self.updateMessage = @"⏳ 正在更新设备信息...";
    [self notifyStatusUpdate];
    
    [DooPushManager.shared updateDeviceInfo];
    
    // 设置超时
    __weak typeof(self) weakSelf = self;
    self.updateTimeoutTimer = [NSTimer scheduledTimerWithTimeInterval:10.0 repeats:NO block:^(NSTimer * _Nonnull timer) {
        if (weakSelf.isUpdatingDevice) {
            weakSelf.isUpdatingDevice = NO;
            weakSelf.updateMessage = nil;
            weakSelf.lastError = @"更新超时";
            [weakSelf notifyStatusUpdate];
        }
    }];
}

- (void)checkPermissionStatus {
    __weak typeof(self) weakSelf = self;
    [DooPushManager.shared checkPushPermissionStatusWithCompletion:^(UNAuthorizationStatus status) {
        dispatch_async(dispatch_get_main_queue(), ^{
            weakSelf.pushPermissionStatus = status;
            [weakSelf updateSDKStatus];
            [weakSelf notifyStatusUpdate];
        });
    }];
}

- (void)clearNotifications {
    [self.notifications removeAllObjects];
    [self notifyStatusUpdate];
}

- (void)reportStatistics {
    [DooPushManager.shared reportStatistics];
    [Logger info:@"📊 手动触发统计数据上报"];
}

#pragma mark - Private Methods

- (void)updateSDKStatus {
    self.deviceToken = [DooPushManager.shared getDeviceToken];
    self.deviceId = [DooPushManager.shared getDeviceId];
    
    if (self.deviceToken && self.deviceToken.length > 0) {
        self.sdkStatus = SDKStatusRegistered;
    } else {
        self.sdkStatus = SDKStatusConfigured;
    }
}

- (void)handleRegistrationSuccess:(NSString *)token {
    self.deviceToken = token;
    self.deviceId = [DooPushManager.shared getDeviceId];
    self.sdkStatus = SDKStatusRegistered;
    self.lastError = nil;
    [self checkPermissionStatus];
    [self notifyStatusUpdate];
}

- (void)handleRegistrationError:(NSError *)error {
    self.sdkStatus = SDKStatusFailed;
    self.lastError = [self userFriendlyMessageForError:error];
    [self checkPermissionStatus];
    [self notifyStatusUpdate];
}

- (void)notifyStatusUpdate {
    // 发送通知给所有监听者
    [[NSNotificationCenter defaultCenter] postNotificationName:DooPushStatusUpdateNotification 
                                                        object:self 
                                                      userInfo:nil];
    
    // 保持向后兼容性
    if (self.statusUpdateCallback) {
        self.statusUpdateCallback();
    }
}

- (NSString *)userFriendlyMessageForError:(NSError *)error {
    // 根据错误类型提供用户友好的消息
    if ([error.domain isEqualToString:NSURLErrorDomain]) {
        switch (error.code) {
            case NSURLErrorNotConnectedToInternet:
            case NSURLErrorNetworkConnectionLost:
                return @"网络连接失败，请检查网络设置";
            case NSURLErrorTimedOut:
                return @"操作超时，请重试";
            case NSURLErrorCancelled:
                return @"操作已取消";
            default:
                return @"网络请求失败，请重试";
        }
    } else if (error.code == 1100) { // DooPushError.pushPermissionDenied
        return @"请在设置中开启推送通知权限";
    } else if (error.code == 1000) { // DooPushError.notConfigured
        return @"SDK未正确配置，请联系开发者";
    } else if (error.code == 1401) { // DooPushError.unauthorized
        return @"应用认证失败，请重新启动应用";
    } else if (error.code == 1500) { // DooPushError.serverError
        return @"服务暂时不可用，请稍后重试";
    }
    
    return error.localizedDescription ?: @"操作失败，请重试";
}

- (NSDictionary *)parseNotificationUserInfo:(NSDictionary *)userInfo {
    NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
    
    // 解析 APNs 格式数据
    NSDictionary *aps = userInfo[@"aps"];
    if (aps) {
        id alert = aps[@"alert"];
        if ([alert isKindOfClass:[NSDictionary class]]) {
            NSDictionary *alertDict = (NSDictionary *)alert;
            result[@"title"] = alertDict[@"title"];
            result[@"content"] = alertDict[@"body"];
        } else if ([alert isKindOfClass:[NSString class]]) {
            result[@"content"] = (NSString *)alert;
        }
    } else {
        // 兼容其他格式
        result[@"title"] = userInfo[@"title"];
        result[@"content"] = userInfo[@"content"] ?: userInfo[@"body"];
    }
    
    // 解析自定义数据
    NSMutableDictionary *customPayload = [[NSMutableDictionary alloc] init];
    for (NSString *key in userInfo.allKeys) {
        if (![key isEqualToString:@"aps"]) {
            customPayload[key] = userInfo[key];
        }
    }
    
    if (customPayload.count > 0) {
        result[@"payload"] = [customPayload copy];
    }
    
    return [result copy];
}

#pragma mark - DooPushDelegate

- (void)dooPush:(DooPushManager *)manager didRegisterWithToken:(NSString *)token {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Logger info:[NSString stringWithFormat:@"🎯 设备注册成功: %@", token]];
        [self handleRegistrationSuccess:token];
    });
}

- (void)dooPush:(DooPushManager *)manager didReceiveNotification:(NSDictionary *)userInfo {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Logger info:[NSString stringWithFormat:@"📱 收到推送通知: %@", userInfo]];
        
        NotificationInfo *notification = [[NotificationInfo alloc] init];
        
        // 解析推送通知数据
        NSDictionary *parsedData = [self parseNotificationUserInfo:userInfo];
        notification.title = parsedData[@"title"];
        notification.content = parsedData[@"content"];
        notification.payload = parsedData[@"payload"];
        
        // 插入到列表开头
        [self.notifications insertObject:notification atIndex:0];
        
        // 限制通知历史数量
        if (self.notifications.count > 50) {
            [self.notifications removeLastObject];
        }
        
        [self notifyStatusUpdate];
    });
}

- (void)dooPush:(DooPushManager *)manager didFailWithError:(NSError *)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Logger error:[NSString stringWithFormat:@"❌ 发生错误: %@", error]];
        
        // 如果正在更新设备信息时出错
        if (self.isUpdatingDevice) {
            [self.updateTimeoutTimer invalidate];
            self.updateTimeoutTimer = nil;
            self.isUpdatingDevice = NO;
            self.updateMessage = nil;
            self.lastError = [NSString stringWithFormat:@"设备信息更新失败: %@", [self userFriendlyMessageForError:error]];
        } else {
            [self handleRegistrationError:error];
        }
        
        [self notifyStatusUpdate];
    });
}

#pragma mark - Optional Delegate Methods

- (void)dooPushDidUpdateDeviceInfo:(DooPushManager *)manager {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Logger info:@"✅ 设备信息更新成功"];
        
        [self.updateTimeoutTimer invalidate];
        self.updateTimeoutTimer = nil;
        
        self.isUpdatingDevice = NO;
        self.updateMessage = @"✅ 设备信息更新成功";
        [self notifyStatusUpdate];
        
        // 3秒后清除更新消息
        __weak typeof(self) weakSelf = self;
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            weakSelf.updateMessage = nil;
            [weakSelf notifyStatusUpdate];
        });
    });
}

- (void)dooPush:(DooPushManager *)manager didChangePermissionStatus:(NSInteger)status {
    dispatch_async(dispatch_get_main_queue(), ^{
        [Logger info:[NSString stringWithFormat:@"🔔 权限状态变更: %ld", (long)status]];
        [self checkPermissionStatus];
    });
}

#pragma mark - Display Text Methods

- (NSString *)displayTextForSDKStatus:(SDKStatus)status {
    switch (status) {
        case SDKStatusNotConfigured:
            return @"未配置";
        case SDKStatusConfigured:
            return @"已配置";
        case SDKStatusRegistering:
            return @"注册中...";
        case SDKStatusRegistered:
            return @"已注册";
        case SDKStatusFailed:
            return @"注册失败";
    }
}

- (NSString *)colorNameForSDKStatus:(SDKStatus)status {
    switch (status) {
        case SDKStatusNotConfigured:
        case SDKStatusFailed:
            return @"red";
        case SDKStatusConfigured:
        case SDKStatusRegistering:
            return @"orange";
        case SDKStatusRegistered:
            return @"green";
    }
}

- (NSString *)displayTextForAuthorizationStatus:(UNAuthorizationStatus)status {
    switch (status) {
        case UNAuthorizationStatusNotDetermined:
            return @"未确定";
        case UNAuthorizationStatusDenied:
            return @"已拒绝";
        case UNAuthorizationStatusAuthorized:
            return @"已授权";
        case UNAuthorizationStatusProvisional:
            return @"临时授权";
        case UNAuthorizationStatusEphemeral:
            return @"短期授权";
        default:
            return @"未知";
    }
}

- (NSString *)colorNameForAuthorizationStatus:(UNAuthorizationStatus)status {
    switch (status) {
        case UNAuthorizationStatusDenied:
            return @"red";
        case UNAuthorizationStatusNotDetermined:
        case UNAuthorizationStatusProvisional:
        case UNAuthorizationStatusEphemeral:
            return @"orange";
        case UNAuthorizationStatusAuthorized:
            return @"green";
        default:
            return @"gray";
    }
}

- (void)dealloc {
    [self.updateTimeoutTimer invalidate];
}

@end

File: sdk/ios/DooPushSDKExample/DooPushSDKExample/PushNotificationManager.swift
//
//  PushNotificationManager.swift
//  DooPushSDKExample
//
//  Created by 韦一 on 2025/8/25.
//

import Foundation
import SwiftUI
import DooPushSDK
import UserNotifications

/// 推送通知管理器 - 实现DooPushDelegate
class PushNotificationManager: NSObject, DooPushDelegate, ObservableObject {
    
    // MARK: - Published Properties
    @Published var sdkStatus: SDKStatus = .notConfigured
    @Published var pushPermissionStatus: UNAuthorizationStatus = .notDetermined
    @Published var deviceToken: String?
    @Published var deviceId: String?
    @Published var lastError: String?
    @Published var notifications: [NotificationInfo] = []
    @Published var isLoading = false
    @Published var isUpdatingDevice = false
    @Published var updateMessage: String?
    
    // MARK: - SDK Status
    enum SDKStatus {
        case notConfigured
        case configured
        case registering
        case registered
        case failed
        
        var displayText: String {
            switch self {
            case .notConfigured:
                return "未配置"
            case .configured:
                return "已配置"
            case .registering:
                return "注册中..."
            case .registered:
                return "已注册"
            case .failed:
                return "注册失败"
            }
        }
        
        var statusColor: Color {
            switch self {
            case .notConfigured, .failed:
                return .red
            case .configured, .registering:
                return .orange
            case .registered:
                return .green
            }
        }
    }
    
    // MARK: - Notification Info
    struct NotificationInfo: Identifiable, Equatable {
        let id = UUID()
        let title: String?
        let content: String?
        let payload: [String: Any]?
        let dedupKey: String?
        let receivedAt: Date
        
        static func == (lhs: NotificationInfo, rhs: NotificationInfo) -> Bool {
            lhs.id == rhs.id
        }
    }
    
    override init() {
        super.init()
        checkPermissionStatus()
        updateSDKStatus()
    }
    
    /// SDK配置完成后检查是否需要自动注册
    func checkAutoRegister() {
        // 检查是否之前已经注册过
        if let _ = DooPushManager.shared.getDeviceToken() {
            // 如果有本地token，自动重新注册以恢复连接
            registerForPushNotifications()
        }
    }
    
    // MARK: - Public Methods
    
    /// 注册推送通知
    func registerForPushNotifications() {
        isLoading = true
        sdkStatus = .registering
        lastError = nil
        
        DooPushManager.shared.registerForPushNotifications { [weak self] token, error in
            DispatchQueue.main.async {
                self?.isLoading = false
                
                if let error = error {
                    self?.handleRegistrationError(error)
                } else if let token = token {
                    self?.handleRegistrationSuccess(token)
                }
            }
        }
    }
    
    /// 更新设备信息
    func updateDeviceInfo() {
        isUpdatingDevice = true
        updateMessage = "正在更新设备信息..."
        
        DooPushManager.shared.updateDeviceInfo()
        
        // 设置超时
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
            if self?.isUpdatingDevice == true {
                self?.isUpdatingDevice = false
                self?.updateMessage = nil
                self?.lastError = "更新超时"
            }
        }
    }
    
    /// 检查权限状态
    func checkPermissionStatus() {
        DooPushManager.shared.checkPushPermissionStatus { [weak self] status in
            DispatchQueue.main.async {
                self?.pushPermissionStatus = status
                self?.updateSDKStatus()
            }
        }
    }
    
    /// 清除通知历史
    func clearNotifications() {
        notifications.removeAll()
    }
    
    /// 手动上报统计数据
    func reportStatistics() {
        DooPushManager.shared.reportStatistics()
        Logger.info("📊 手动触发统计数据上报")
    }
    
    // MARK: - Private Methods
    
    private func updateSDKStatus() {
        deviceToken = DooPushManager.shared.getDeviceToken()
        deviceId = DooPushManager.shared.getDeviceId()
        
        if deviceToken != nil {
            sdkStatus = .registered
        } else {
            sdkStatus = .configured
        }
    }
    
    private func handleRegistrationSuccess(_ token: String) {
        deviceToken = token
        deviceId = DooPushManager.shared.getDeviceId()
        sdkStatus = .registered
        lastError = nil
        checkPermissionStatus()
    }
    
    private func handleRegistrationError(_ error: Error) {
        sdkStatus = .failed
        lastError = DooPushErrorHandler.userFriendlyMessage(for: error)
        checkPermissionStatus()
    }
    
    // MARK: - DooPushDelegate
    
    func dooPush(_ manager: DooPushManager, didRegisterWithToken token: String) {
        DispatchQueue.main.async {
            Logger.info("🎯 设备注册成功: \(token)")
            self.handleRegistrationSuccess(token)
        }
    }
    
    func dooPush(_ manager: DooPushManager, didReceiveNotification userInfo: [AnyHashable: Any]) {
        DispatchQueue.main.async {
            Logger.info("📱 收到推送通知: \(userInfo)")
            
            let parser = DooPushNotificationParser.parse(userInfo)
            let incoming = NotificationInfo(
                title: parser.title,
                content: parser.content,
                payload: parser.payload,
                dedupKey: parser.dedupKey,
                receivedAt: Date()
            )
            
            // 使用 dedupKey 去重
            if let key = incoming.dedupKey {
                let exists = self.notifications.contains { $0.dedupKey == key }
                if !exists {
                    self.notifications.insert(incoming, at: 0)
                }
            } else {
                // 无 dedupKey 时不做去重
                self.notifications.insert(incoming, at: 0)
            }
            
            // 限制通知历史数量
            if self.notifications.count > 50 {
                self.notifications.removeLast()
            }
        }
    }
    
    func dooPush(_ manager: DooPushManager, didFailWithError error: Error) {
        DispatchQueue.main.async {
            Logger.error("❌ 发生错误: \(error)")
            
            // 如果正在更新设备信息时出错
            if self.isUpdatingDevice {
                self.isUpdatingDevice = false
                self.updateMessage = nil
                self.lastError = "设备信息更新失败: \(DooPushErrorHandler.userFriendlyMessage(for: error))"
            } else {
                self.handleRegistrationError(error)
            }
        }
    }
    
    // MARK: - Optional Delegate Methods
    
    func dooPushDidUpdateDeviceInfo(_ manager: DooPushManager) {
        DispatchQueue.main.async {
            Logger.info("✅ 设备信息更新成功")
            self.isUpdatingDevice = false
            self.updateMessage = "设备信息更新成功"
            
            // 3秒后清除更新消息
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
                self?.updateMessage = nil
            }
        }
    }
    
    func dooPush(_ manager: DooPushManager, didChangePermissionStatus status: Int) {
        DispatchQueue.main.async {
            Logger.info("🔔 权限状态变更: \(status)")
            self.checkPermissionStatus()
        }
    }
}

// MARK: - Extensions

extension UNAuthorizationStatus {
    var displayText: String {
        switch self {
        case .notDetermined:
            return "未确定"
        case .denied:
            return "已拒绝"
        case .authorized:
            return "已授权"
        case .provisional:
            return "临时授权"
        case .ephemeral:
            return "短期授权"
        @unknown default:
            return "未知"
        }
    }
    
    var statusColor: Color {
        switch self {
        case .denied:
            return .red
        case .notDetermined, .provisional, .ephemeral:
            return .orange
        case .authorized:
            return .green
        @unknown default:
            return .gray
        }
    }
}
